extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,375 @@
+      // the external CID.
+      // The dialogPromise_ is never rejected,
+      // the user only really has 1 option to accept/dismiss (to resolve)
+      // the notification or have the nagging notification sitting there
+      // (to never resolve).
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getAsyncCid_()
+        .then(this.postShowEndpoint_.bind(this))
+        .then(this.onPostShowEndpointSuccess_.bind(this));
+  }
+
+  /** @override */
+  show() {
+    this.element.style.display = '';
+    this.element.classList.add('amp-active');
+    return this.dialogPromise_;
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_();
+    this.postDismissEnpoint_();
+  }
+}
+
+
+/**
+ * UserNotificationManager handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationManager {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDeferDef>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
+
+    let documentReadyPromise_ = new Promise((resolve) => {
+      this.documentReadyResolve_ = resolve;
+    });
+
+    // TODO(erwinm, https://github.com/ampproject/amphtml/issues/1226)
+    // Add `whenDocumentReady` function in document-state
+    // that returns a promise.
+    onDocumentReady(this.win_.document, this.documentReadyResolve_);
+
+    /** @private {!Promise} */
+    this.nextInQueue_ = this.viewer_.whenVisible().then(() => {
+      // we block on both viewer visible and when dom ready.
+      return documentReadyPromise_;
+    });
+  }
+
+  /**
+   * Retrieve a promise associated to an `amp-user-notification` component
+   * that is resolved when user agrees to the terms.
+   * @param {string} id
+   * @return {!Promise<string>}
+   */
+  get(id) {
+    return this.getElementDeferById_(id).promise;
+  }
+
+  /**
+   * Register an instance of `amp-user-notification`.
+   * @param {string} id
+   * @param {!UserNotification} userNotification
+   * @package
+   */
+  registerUserNotification(id, userNotification) {
+    const deferred = this.getElementDeferById_(id);
+
+    // Compose the registered notifications into a promise queue
+    // that blocks until one notification is dismissed.
+    this.nextInQueue_ = this.nextInQueue_
+        .then(() => {
+          if (this.win_.document.getElementById(id) != null) {
```

Comment:
I think it's good, except for two things:

1. It should be `== null` above. Right?
2. Does it make sense to still try with `shouldShow`? I think we can just skip this instance and go straight to the next one. E.g. in this if we can return `Promise.reject` or something like that.